File: 0155-min-stack/0155-min-stack.java
class MinStack {
private Stack<Integer>st;
private Stack<Integer>min;
public MinStack() {
        st=new Stack<>();
        min=new Stack<>();
    }
    
    void push(int val) {
        if(st.size()==0 ||min.peek()>=val){
            min.push(val);
        }
        st.push(val);
    }

    
    void pop() {
    int ele1=st.pop();    
    int ele2=min.peek();
    if (ele1==ele2){
        min.pop();
    }
    }
    
    int top() {
        return st.peek();
    }
    
    int getMin() {
        return min.peek();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
public:
stack<long long>st; long long mini=INT_MAX; 
    MinStack() {
        
    }
    
    void push(int value) {
       long long val=(long long)value;
        if(st.empty()){
            mini=val;
            st.push(val);
        }
        else{
            if(val>mini) st.push(val);
            else{
                st.push(2*val*1LL-mini);
                mini=val;
            }
        }
    }
    
    void pop() {
        
        long long n=st.top();
        st.pop();
        if(n<mini){
            mini=2*mini-n;
        }
    }
    
    int top() {
        if(st.empty()) return 0;
        long long n=st.top();
        if(mini<n) return n;
        return mini;
    }
    
    int getMin() {
        return mini;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */